### PR TITLE
Soft delete comments so that entire thread is not deleted

### PR DESCRIPTION
**Pull request by alcalyn**
Related to #507 

Currently, when deleting a comment that have replies, the comment and all the replies were removed.

This PR only soft deleted comments so that replies stay visible.

Screenshot_2019-11-25 date-toggle-demo

- [x] Back: Update tests to assert comment is still there, but marked as deleted
- [x] Back: Update server to soft delete comment instead of destroy
- [x] Back: Update API to no longer returns comment text and account if soft deleted
- [x] Back: Migration to make `comment.accountId` optional and add the `deletedAt` column
- [x] Federation: Replace deleted comments with a `Tombstone` object
- [x] Federation: Handle comment deletion federation requests by marking comments as deleted
- [x] Federation: Handle deleted comments when following a new instance and fetching all comments for the first time
- [x] Front: Display deleted comment as deleted on page load
- [x] Front: Display deleted comment as deleted dynamically when clicking on Delete

### CODE REVIEW

**Comment by alcalyn** [comment]
I'm having trouble with:

- `comment.totalReplies` which is still either `0` or `undefined`
- I don't know how to run the migration I created (tried to restart server, to `npm run build:server`...)

**Comment by rigelk** [comment]
> I don't know how to run the migration I created (tried to restart server, to `npm run build:server`...)

Have you incremented LAST_MIGRATION_VERSION in `server/initializers/constants.ts`?

**Comment by alcalyn** [comment]
Ha, didn't aware of this constant. Just incremented it, run `npm run build:server` and it worked! Thank you

File: server/models/video/video-comment.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { AllowNull, BelongsTo, Column, CreatedAt, DataType, ForeignKey, Is, Model, Scopes, Table, UpdatedAt } from 'sequelize-typescript'
+import { AllowNull, BelongsTo, Column, CreatedAt, DataType, DeletedAt, ForeignKey, Is, Model, Scopes, Table, UpdatedAt } from 'sequelize-typescript'
```
Comment (LeoMouyna) [inline server/models/video/video-comment.ts]: @alcalyn This line seems too long according to travis :upside_down_face:

File: client/src/app/videos/+video-watch/comment/video-comment.model.ts [hunk 2]
```diff
@@ -28,6 +30,8 @@ export class VideoComment implements VideoCommentServerModel {
     this.videoId = hash.videoId
     this.createdAt = new Date(hash.createdAt.toString())
     this.updatedAt = new Date(hash.updatedAt.toString())
+    this.deletedAt = null === hash.deletedAt ? null : new Date(hash.deletedAt.toString())
```
Comment (LeoMouyna) [inline client/src/app/videos/+video-watch/comment/video-comment.model.ts]: nitpick: I'm not sure this equal checker is need.
'has.deletedAt ? new Date(hash.deletedAt.toString()) : null' should do the same.

Comment (alcalyn) [reply client/src/app/videos/+video-watch/comment/video-comment.model.ts]: thanks for the review :)

**Comment by Chocobozzz** [comment]
@alcalyn Did you manage to fix your `totalReplies` issue?

I think it would be simpler to not use the paranoid mode and instead manually update the `deletedAt` field, because we also need to remove the `text` field (set to null or empty string).

We also have to deal with federation:
 * Add a `deleted` field and use a `Tombstone` object: https://www.w3.org/TR/activitypub/#delete-activity-outbox
 * Correctly process this `deleted` field

**Comment by alcalyn** [comment]
Hello, yes I could add the scope `ScopeNames.ATTRIBUTES_FOR_API` but it works only for top-level comments. I couldn't have replies count for replies.

I was not sure if we should also erase comment from db. I'll try to do it without paranoid.

Same for account that posted the comment: if we also delete it, we have to make it optionnal, and make many changes

**Comment by Chocobozzz** [comment]
> Same for account that posted the comment: if we also delete it, we have to make it optionnal, and make many changes

I don't think we have the choice. Users expects that we don't keep this information if they delete their comment. If you did not plan to make these changes, we can merge your work in a WIP branch and I'll continue it when I'll have time :)

**Comment by alcalyn** [comment]
I may have time to finish it, I'll redo it without `paranoid`

**Comment by alcalyn** [comment]
**I get issues** when running them, it looks like the migration I created is not run on testing database.

I'm running tests with:

``` bash
$ npm run build -- --light
$ npm run clean:server:test

$ npm run mocha -- --exit --require ts-node/register/type-check --bail server/tests/api/videos/video-comments.ts 

> peertube@2.0.0 mocha /var/www/PeerTube
> mocha "--exit" "--require" "ts-node/register/type-check" "--bail" "server/tests/api/videos/video-comments.ts"

  Test video comments
    ✓ Should not have threads on this video (42ms)
    ✓ Should create a thread in this video (141ms)
    ✓ Should list threads of this video (84ms)
    ✓ Should get all the thread created (38ms)
    ✓ Should create multiple replies in this thread (467ms)
    ✓ Should get correctly the replies (52ms)
    ✓ Should create other threads (193ms)
    ✓ Should list the threads (44ms)
    1) Should delete a reply

  8 passing (19s)
  1 failing

  1) Test video comments
       Should delete a reply:
     Error: expected 204 "No Content", got 500 "Internal Server Error"
```

The error is `null value in column "accountId" violates not-null constraint`:

```
$ tail -n 2 test1/logs/peertube.log 
{"err":"SequelizeDatabaseError: null value in column \"accountId\" violates not-null constraint [...] }
{"message":"127.0.0.1 - - [23/Nov/2019:13:26:37 +0000] \"DELETE /api/v1/videos/1/comments/4 HTTP/1.1\" 500 - \"-\" \"node-superagent/3.8.3\"\n","level":"info","label":"localhost:9001","timestamp":"2019-11-23T13:26:37.059Z"}
```

I think I have to run migrations on test database but I don't know how to.

### About Tombstone object

The api now returns a Tombstone for deleted comments:

**before deletion**:
```
$ curl -X GET http://0.0.0.0:9000/videos/watch/e7c647ea-a9bc-4eee-99b2-82109ec70a5c/comments/17 -H "accept: application/activity+json" | jq

{
  "type": "Note",
  "id": "http://localhost:9000/videos/watch/e7c647ea-a9bc-4eee-99b2-82109ec70a5c/comments/17",
  "content": "Hello",
  "inReplyTo": "http://localhost:9000/videos/watch/e7c647ea-a9bc-4eee-99b2-82109ec70a5c",
  "updated": "2019-11-23T13:14:39.228Z",
  "published": "2019-11-23T13:14:39.199Z",
  "url": "http://localhost:9000/videos/watch/e7c647ea-a9bc-4eee-99b2-82109ec70a5c/comments/17",
  "attributedTo": "http://localhost:9000/accounts/root",
  "tag": [],
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ],
  "cc": [
    "http://localhost:9000/accounts/root/followers"
  ],
  "@context": ["..."]
}
```

**after deletion**:
```
$ curl -X GET http://0.0.0.0:9000/videos/watch/e7c647ea-a9bc-4eee-99b2-82109ec70a5c/comments/17 -H "accept: application/activity+json" | jq

{
  "@context": ["..."],
  "id": "http://localhost:9000/videos/watch/e7c647ea-a9bc-4eee-99b2-82109ec70a5c/comments/17",
  "type": "Tombstone",
  "formerType": "Note",
  "published": "2019-11-23T13:14:39.199Z",
  "updated": "2019-11-23T13:14:55.596Z",
  "deleted": "2019-11-23T13:14:55.592Z"
}
```

**Comment by Chocobozzz** [comment]
You need to allow `null`  [here](https://github.com/Chocobozzz/PeerTube/blob/develop/server/models/video/video-comment.ts#L180) and [here](https://github.com/Chocobozzz/PeerTube/blob/develop/server/models/account/account.ts#L204)

**Comment by alcalyn** [comment]
Thanks it worked, migrations was actually run. I'm working on others tests about federation.

**Comment by alcalyn** [comment]
Should be ready to review

**Comment by Chocobozzz** [comment]
Great job @alcalyn! Just one thing: you'll have to handle Tombstone objects in [this file](https://github.com/Chocobozzz/PeerTube/blob/develop/server/lib/activitypub/video-comments.ts), so that when you follow a new server and crawl its outbox, you create the appropriate tree. Without that, you'll miss some comment trees because the remote body won't be considered [as valid](https://github.com/Chocobozzz/PeerTube/blob/develop/server/lib/activitypub/video-comments.ts#L129)

**Comment by alcalyn** [comment]
Ok I didn't tried this use case yet. I'm adding tests and fix it

**Comment by alcalyn** [comment]
I fixed crawl of deleted comments when subscribing a new instance in this commit: https://github.com/Chocobozzz/PeerTube/pull/2165/commits/b1f7c4f9d3222a116b5d8305131653651e23c364

**Comment by Chocobozzz** [comment]
Thanks @alcalyn :+1:

**Pull request merged**